off-chain/src/Network/Users.purs

```diff
@@ -135,20 +151,140 @@ startUserNode
   => UserConfiguration a
   -> Aff (Fiber Unit /\ UserNode a)
 startUserNode conf = do
+  node <- newUserNode conf
+  fiber <- startActionStatusCheck node
+  pure $ fiber /\ node
+
+newUserState :: forall a. Aff (UserState a)
+newUserState = do
+  actionsSentQueue <- liftEffect $ Queue.new
   actionsSent <- liftEffect $ Ref.new OrderedMap.empty
-  transactionsSent <- liftEffect $ Ref.new OrderedMap.empty
-  submitedTransactions <- liftEffect $ Ref.new OrderedMap.empty
+  resultsQueue <- liftEffect $ Queue.new
+  pure $ wrap
+    { actionsSentQueue
+    , actionsSent
+    , resultsQueue
+    }
+
+singTransactionAndSend
+  :: forall a
+   . Show a
+  => UserNode a
+  -> { uuid :: UUID
+     , txCborHex :: String
+     , action :: UserAction a
+     , previousStatus :: ActionStatus
+     }
+  -> Aff (Maybe String)
+singTransactionAndSend userNode afs = do
+  tx <- TxHex.fromCborHex afs.txCborHex
+  signedTx <- signTx userNode tx
+  signedCbor <- TxHex.toCborHex signedTx
+
   let
-    node = UserNode
-      { state: UserState
-          { actionsSent
-          , transactionsSent
-          , submitedTransactions
-          }
+    singedRequest = SendSignedTransaction
+      { uuid: afs.uuid, txCborHex: signedCbor }
+  result <- (getUserHandlers userNode).sendSignedToLeader singedRequest
+  case result of
+    Right _ -> do
+      log $ "User: Successfully signed and sent Tx " <> show afs.uuid
+        <> " to the Leader"
+      pure Nothing
+    Left e -> pure <<< Just $ show e
+
+{- TODO: Many possible errors can be caught here:
+ - error during status request to the leader over HTTP (e.g. leader is offline)
+ - can't parse UUID returned by the leader
+ - JSON decode errors
+ - error during encoding and deconding transaction CBOR
+ - error while running signing Contract
+ - error while sending signed Tx to the leader over HTTP
+ - AcceptSignedTransactionError
+
+ Not sure atm how granular error handler need to be.
+ For now, we put a string with the error.
+-}
+-- | This is intended to be the only function that manipulates `actionsSent` in the 
+-- | `UserCOnfiguration`
+makeActionsSentHandler
+  :: forall a
+   . Show a
+  => UserNode a
+  -> { uuid :: UUID
+     , action :: UserAction a
+     , status :: ActionStatus
+     , previousStatus :: ActionStatus
+     }
+  -> Effect Unit
+makeActionsSentHandler userNode record = launchAff_ $
+  case record.status of
+    -- This must block the thread, otherwise if we allow `actionHandler` to spawn
+    -- threads, we would face race conditions again.
+    AskForSignature afs -> do
+      result <- singTransactionAndSend userNode
```

Comment:
`singTransactionAndSend` can throw an error. At least from HTTP side with current handlers. To/From CBOR conversions if I'm not mistaken also throw errors.